lib/auth/webauthn/messages_test.go

```diff
@@ -56,3 +58,252 @@ func TestCredentialAssertionResponse_json(t *testing.T) {
+				cp.Response.User.DisplayName = ""
+				return &cp
+			},
+			wantErr: "user display name",
+		},
+		{
+			name: "rrk nil user ID",
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.ID = nil
+				return &cp
+			},
+			wantErr: "user ID",
+		},
+		{
+			name:           "cc without rrk but pass alwaysCreateRK",
+			createCC:       func() *wanlib.CredentialCreation { return okCC },
+			alwaysCreateRK: true,
+			wantErr:        "relying party name required",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.createCC().Validate(test.alwaysCreateRK)
+			if test.wantErr != "" {
+				require.Error(t, err, "Validate returned err = nil, want %q", test.wantErr)
+				assert.Contains(t, err.Error(), test.wantErr, "Validate returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestCredentialAssertion_Validate(t *testing.T) {
+	okAssertion := &wanlib.CredentialAssertion{
+		Response: protocol.PublicKeyCredentialRequestOptions{
+			Challenge:      make([]byte, 32),
+			RelyingPartyID: "example.com",
+			AllowedCredentials: []protocol.CredentialDescriptor{
+				{Type: protocol.PublicKeyCredentialType, CredentialID: []byte{1, 2, 3, 4, 5}},
+			},
+		},
+	}
+
+	nilChallengeAssertion := *okAssertion
+	nilChallengeAssertion.Response.Challenge = nil
+
+	emptyRPIDAssertion := *okAssertion
+	emptyRPIDAssertion.Response.RelyingPartyID = ""
+	tests := []struct {
+		name      string
+		assertion *wanlib.CredentialAssertion
+		wantErr   string
+	}{
+		{
+			name:      "ok", // check that good params are good
+			assertion: okAssertion,
+			wantErr:   "",
+		},
+		{
+			name:    "nil assertion",
+			wantErr: "assertion required",
+		},
+		{
+			name:      "assertion without challenge",
+			assertion: &nilChallengeAssertion,
+			wantErr:   "challenge",
+		},
+		{
+			name:      "assertion without RPID",
+			assertion: &emptyRPIDAssertion,
+			wantErr:   "relying party ID",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.assertion.Validate()
+			if test.wantErr != "" {
+				require.Error(t, err, "Validate returned err = nil, want %q", test.wantErr)
+				assert.Contains(t, err.Error(), test.wantErr, "Validate returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestRequireResidentKey(t *testing.T) {
+	tests := []struct {
+		name string
+		in   protocol.AuthenticatorSelection
+		want bool
+	}{
+		{
+			name: "nothing set",
+			in:   protocol.AuthenticatorSelection{},
+			want: false,
+		},
+		{
+			name: "prefer ResidentKey over RequireResidentKey",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementDiscouraged,
+				RequireResidentKey: protocol.ResidentKeyRequired(),
```

Comment:
This is should cause an error - `RequireResidentKey` can only be true if `ResidentKey` is "required" or unknown/absent. (At least that's my understanding of the spec.)

Good call in adding this test case.

(from `ResidentKey`) "but client platforms MUST ignore unknown values, treating an unknown value as if the member does not exist." (it then falls back to `RequireResidentKey`)

(from `RequireResidentKey`) "Relying Parties SHOULD set it to true if, and only if, residentKey is set to required."


https://www.w3.org/TR/webauthn-2/#dictionary-authenticatorSelection

Reply:
done, it makes `ResidentKeyRequired` a little bit more complex but now we are also returning error there.